SelectionDAG store merging default implementations are unreasonable #90714

Devesh-Yadav10 opened the pull request:
### Summary
The default implementation of `mergeStoresAfterLegalization` in `TargetLowering` unconditionally returned `true`, which could cause infinite loops when a backend decomposes stores during legalization. 

This patch updates the default behavior to return `isOperationLegal(ISD::STORE, MemVT)`, ensuring that merging only happens when storing `MemVT` is explicitly legal. This provides a more conservative and safer default.

### Impact
- Prevents infinite loops in custom store legalization.  
- Some targets that previously relied on the unconditional `true` may need to override this function explicitly.  
- A manageable number of test cases break, which suggests that some backends need updates to handle this more correctly.

### Next Steps
If necessary, the API may need to be extended to consider both register and memory value types for more precise store merging decisions.

Devesh-Yadav10 committed:
Updated unconditional true in mergeStoresAfterLegalization to checking isOperationLegal(ISD::STORE, MemVT)

nikic reviewed (commented):
Please update the failing tests.

nikic commented:
Though from a very cursory look, I doubt this makes sense. This predicate just checks whether store merging is attempted at all, there are more checks (isTypeLegal, canMergeStoresTo, allowsMemoryAccess) before actually doing the merge. Possibly your target is not handling some of these correctly?

Devesh-Yadav10 commented:
> Though from a very cursory look, I doubt this makes sense. This predicate just checks whether store merging is attempted at all, there are more checks (isTypeLegal, canMergeStoresTo, allowsMemoryAccess) before actually doing the merge. Possibly your target is not handling some of these correctly?

I see your point about additional checks. My concern was that the default implementation allows merging too aggressively, leading to infinite loops when stores are decomposed. Do you think making canMergeStoresTo more conservative would be a better approach?

Devesh-Yadav10 committed:
[GlobalISel] Introduce G_POISON to distinguish poison from undef

This patch adds G_POISON as a new opcode in GlobalISel to represent poison values separately from G_IMPLICIT_DEF. This distinction is necessary to mirror LLVM IR semantics more accurately and enable better optimizations, particularly when legalization introduces padding vector elements.

Devesh-Yadav10 committed:
[GlobalISel] Handle G_POISON in isLegalOrBeforeLegalizer

This patch ensures that G_POISON is treated similarly to G_IMPLICIT_DEF in isLegalOrBeforeLegalizer. 
This change is necessary to correctly distinguish poison values from undef in GlobalISel.

Devesh-Yadav10 committed:
[GlobalISel] Add support for G_POISON in MachineIRBuilder

This patch introduces G_POISON as a distinct opcode from G_IMPLICIT_DEF in
MachineIRBuilder, allowing better handling of poison values in GlobalISel.

Devesh-Yadav10 committed:
[X86][GlobalISel] Add support for G_POISON in instruction selection

This patch updates the X86 GlobalISel instruction selector to handle G_POISON
alongside G_IMPLICIT_DEF. The changes ensure proper instruction selection and
validation for G_POISON.

- Added case for G_POISON in instruction selection.
- Updated assertions to include G_POISON.
- Ensured all relevant checks now consider G_POISON.

arsenm commented on llvm/lib/CodeGen/GlobalISel/MachineIRBuilder.cpp:
```diff
@@ -642,6 +642,11 @@ MachineInstrBuilder MachineIRBuilder::buildUndef(const DstOp &Res) {
   return buildInstr(TargetOpcode::G_IMPLICIT_DEF, {Res}, {});
 }
 
+MachineInstrBuilder MachineIRBuilder::buildPoison(const DstOp &Res) {
```
Unrelated change

Devesh-Yadav10 committed:
[GlobalISel] Extend MIPatternMatch to support G_POISON

Devesh-Yadav10 committed:
[AArch64][GISel] Add G_POISON to LegalizerInfo

Devesh-Yadav10 committed:
[AArch64][GISel] Extend Instruction Selection to Handle G_POISON

Add support for G_POISON in AArch64 instruction selection, treating it similarly 
to G_IMPLICIT_DEF. This ensures that poisoned values are correctly recognized 
and handled during GlobalISel processing.

Devesh-Yadav10 committed:
[GlobalISel] Add G_POISON Mapping in SelectionDAGCompat.td

Extend SelectionDAG compatibility by adding G_POISON alongside G_IMPLICIT_DEF. 
This ensures correct handling of poison values in GlobalISel transformations.

Devesh-Yadav10 committed:
[GlobalISel] Include G_POISON in LostDebugLocObserver

Extend LostDebugLocObserver to track G_POISON, ensuring consistent handling 
of poison values in debug location tracking.

Devesh-Yadav10 committed:
[GlobalISel] Add G_POISON support in CombinerHelperVectorOps

Extend vector operation legalization to handle G_POISON alongside G_IMPLICIT_DEF.
This ensures poison values are properly considered during vector transformations.

Devesh-Yadav10 committed:
[GlobalISel] Add test coverage for G_POISON in LegalizerHelperTest

Extended test cases to verify G_POISON handling alongside G_IMPLICIT_DEF.
This ensures correctness of legalization behavior for poison values.

Devesh-Yadav10 committed:
[RISCV][GlobalISel] Add handling for G_POISON in RegisterBankInfo

Extended register bank handling to include G_POISON alongside G_IMPLICIT_DEF.
This improves correctness in instruction selection and register assignment.

Devesh-Yadav10 committed:
[AArch64][GlobalISel] Extend G_IMPLICIT_DEF handling to include G_POISON

Modified AArch64PostLegalizerLowering to treat G_POISON equivalently to G_IMPLICIT_DEF.
This ensures correct handling of undefined values in instruction legalization.

Devesh-Yadav10 committed:
[GlobalISel][CSE] Extend tests to include G_POISON alongside G_IMPLICIT_DEF

Updated CSETest.cpp to check for G_POISON, ensuring proper handling of undefined values
in GlobalISel's common subexpression elimination.

Devesh-Yadav10 committed:
[GlobalISel] Replace G_IMPLICIT_DEF with G_POISON in Combine.td

Devesh-Yadav10 committed:
Merge branch 'main' into main
